src/ReactiveDomain.Core/Metadata.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace ReactiveDomain
+{
+	public class Metadata : IMetadata
+	{
+		private readonly JsonSerializer _serializer;
+		private readonly Dictionary<string, object> _cache;
+
+		public Metadata() 
+		{
+			_cache = new Dictionary<string, object>();
+		}
+
+		public Metadata(JObject root, JsonSerializer serializer)
+		{
+			_cache = new Dictionary<string, object>();
+			Root = root;
+			_serializer = serializer;
+		}
+
+		public T Read<T>()
+		{
+			if (!TryRead<T>(out var value)) throw new MetadatumNotFoundException();
+			return value;
+		}
+
+		public bool TryRead<T>(out T value)
+		{
+			if (_cache.TryGetValue(typeof(T).Name, out var cached))
+			{
+				value = (T)cached;
+				return true;
+			}
+
+			if (Root != null && Root.TryGetValue(typeof(T).Name, out var token))
+			{
+				value = token.ToObject<T>(_serializer);
+				_cache[typeof(T).Name] = value;
+				return true;
+			}
+
+			value = default;
+			return false;
+		}
+
+		public void Write<T>(T metadatum)
+		{
+			_cache[typeof(T).Name] = metadatum;
+		}
+
+		internal IReadOnlyDictionary<string, object> GetData()
+		{
+			return _cache;
```

Comment:
Before returning `_cache`, we should deserialize from `Root` into `_cache`, while being careful not to overwrite anything that's already in the cache since that could overwrite things that were set by calls to `Write()`.